{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eef31703",
   "metadata": {},
   "source": [
    "## Notebook for making figures made with R in 'paper name'\n",
    "Anneliek ter Horst\n",
    "\n",
    "- Statistical tests\n",
    "- PCOA plot\n",
    "- Upset Plot\n",
    "- Bar plot \n",
    "- Genome figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f5480a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries needed\n",
    "library(UpSetR)\n",
    "library(tidyverse)\n",
    "library(vegan)\n",
    "library(ggplot2)\n",
    "library(ape)\n",
    "library(ggrepel)\n",
    "library(gggenes)\n",
    "library(RColorBrewer)\n",
    "library(scales)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3079bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCOA plot\n",
    "\n",
    "# load data, coverage table\n",
    "otu <- read.table(\"normalized_coverage_table.csv\", \n",
    "              sep=\",\", header=T, row.names=1)\n",
    "\n",
    "# load data, metadata\n",
    "cat_meta <- read.table(\"metadata_trees.csv\", \n",
    "                         sep=\",\", header=T, row.names=1)\n",
    "\n",
    "# Transpose dataframe\n",
    "otu <- t(otu)\n",
    "\n",
    "# remove rows with all zeros\n",
    "otu <- otu[apply(otu[,-1], 1, function(x) !all(x==0)),]\n",
    "\n",
    "# transform the data\n",
    "otu.xform <- decostand(otu, method=\"hellinger\")\n",
    "\n",
    "# create dissimilarity matrix for the vOTU table with bray method\n",
    "otu.dist <- as.matrix(vegdist(otu.xform, method='bray'))\n",
    "\n",
    "# nmds\n",
    "otu.nmds <- metaMDS(otu.dist)\n",
    "otu.nmds$stress\n",
    "\n",
    "# perform pcoa with ape package pcoa\n",
    "pcoa <- pcoa(as.dist(otu.dist))\n",
    "\n",
    "# Bind the metadata and dissimilarity data based on the sample names\n",
    "cat_meta.ordered <- cat_meta[match(row.names(otu.nmds$points), row.names(cat_meta)),] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb8d23be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a dataframe named axes, put pcoa values in there\n",
    "axes <- as.data.frame(pcoa$vectors)\n",
    "\n",
    "# calculate the eigenvalues for each pcoa axes \n",
    "eigval <- round(pcoa$values$Relative_eig * 100, digits = 2)\n",
    "\n",
    "# Put eigenvalues in their own df\n",
    "eigval <- data.frame( PC = 1:length(eigval), Eigval = eigval)\n",
    "\n",
    "axes$SampleID <- rownames(axes)\n",
    "\n",
    "cat_meta.ordered$SampleID <- rownames(cat_meta.ordered)\n",
    "\n",
    "# perform pcoa with ape package pcoa\n",
    "pcoa <- pcoa(as.dist(otu.dist))\n",
    "\n",
    "# make a dataframe named axes, put pcoa values in there\n",
    "axes <- as.data.frame(pcoa$vectors)\n",
    "\n",
    "# Give df extra column with the rownames in it \n",
    "axes$SampleID <- rownames(axes)\n",
    "\n",
    "cat_meta.ordered$SampleID <- rownames(cat_meta.ordered)\n",
    "\n",
    "axes <- merge(cat_meta.ordered, axes, by.x = \"SampleID\", by.y = \"SampleID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f6367c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# permanova on viral distribution vs tree family\n",
    "pmanova2 = adonis(as.dist(otu.dist) ~ sci_fam, data = cat_meta.ordered) \n",
    "pmanova2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45197ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create pcoa plot\n",
    "pdf(\"PCOA_colored_tree_fam.pdf\")\n",
    "p <- ggplot(axes, aes(Axis.1, Axis.2)) \n",
    "a <- p + geom_point(aes(colour=sci_fam), size=4.5 ,alpha=0.8) + \n",
    "geom_text_repel(aes(label=sci_name_short, fontface = \"italic\", size=17)) +\n",
    "    xlab(paste(\"PCo1 (\", eigval$Eigval[1], \" %)\", sep = \"\")) +\n",
    "    ylab(paste(\"PCo2 (\", eigval$Eigval[2], \" %)\", sep = \"\")) +\n",
    "  #scale_color_brewer(name = \"sci_fam\", palette = 'Dark2') +\n",
    "  #scale_color_gradientn(colors = RColorBrewer::brewer.pal(11, \"Spectral\")) +\n",
    "    scale_color_manual(values = c(\"#003f5c\", \"#bc5090\", \"#ffa600\")) +\n",
    "    geom_vline(xintercept = 0, linetype = 2) +\n",
    "    geom_hline(yintercept = 0, linetype = 2) +\n",
    "    theme_bw() +\n",
    "    theme(text = element_text(size = 18),\n",
    "        legend.position = \"bottom\") +\n",
    "    guides(colour = guide_legend(title.position = \"top\", title.hjust = 0.5)) \n",
    "print(a)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3a7ac13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upset plot\n",
    "# Load data and format\n",
    "\n",
    "# metadata\n",
    "map <- read.csv(\"metadata_trees.csv\", header = T)\n",
    "map <- map %>% \n",
    "  rename(\"SampleID\" = \"HTS_ID\")\n",
    "\n",
    "# load data, coverage table\n",
    "otu <- read.table(\"normalized_coverage_table.csv\", \n",
    "              sep=\",\", header=T, row.names=1)\n",
    "\n",
    "# clean the coverage table dataframe\n",
    "row.names(otu) <- otu$contig\n",
    "otu <- otu[,-1]\n",
    "otu <- otu[,colnames(otu) %in% map$SampleID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d09cb392",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create UpSet plot (Sorry i know this is not pretty)\n",
    "pdf(\"UpSetplot.pdf\")\n",
    "upset(fromList(list(Cupressaceae = filter(site.occurrence, SampleID == \"RW3\" | SampleID == \"RW4\" |\n",
    "                                         SampleID == \"Cyp4\" | SampleID == \"Cyp3\" | SampleID == \"Cedar2\")$OTU_ID,\n",
    "                                    Fagaceae = filter(site.occurrence, SampleID == \"Oak9\" | SampleID == \"Oak8\" |\n",
    "                                                     SampleID == \"Oak7\" | SampleID == \"Oak11\" | SampleID == \"Oak10\" | SampleID == \"Oak12\")$OTU_ID,\n",
    "                                    Pinaceae = filter(site.occurrence, SampleID == \"Pine6\" | SampleID == \"Pine5\" |\n",
    "                                                     SampleID == \"Pine4\" | SampleID == \"Fir3\" | SampleID == \"Fir4\" )$OTU_ID)), \n",
    "      text.scale=2, nsets = 16, nintersects = 36, order.by = \"freq\") \n",
    "                    #nintersects = NA)))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d8c6b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genome figures\n",
    "# Open the file with the genomes\n",
    "genome <- read.table(\"contigs_1kb_3prot.csv\",\n",
    "                   sep=\",\", header=T)\n",
    "\n",
    "\n",
    "# add gene direction\n",
    "genome$direction <- ifelse(genome$strand == \"forward\", 1, -1)\n",
    "\n",
    "\n",
    "# Plot the genomes in arrows \n",
    "p <- ggplot2::ggplot(genome, ggplot2::aes(xmin = start, xmax = end, y =\n",
    "                                            molecule, fill = gene, forward = direction)) +\n",
    "    geom_gene_arrow() \n",
    "\n",
    "# Give my color scheme instead\n",
    "p2 <- p + scale_fill_manual(values = mycolors) + theme(text= element_blank(),\n",
    "        axis.text = element_text(size = 10))\n",
    "\n",
    "# Facet them so that each of them has its own bar (it goes wrong)\n",
    "p3 <- p2 + ggplot2::facet_wrap(~ molecule, scales = \"free\", ncol = 1) + theme_genes() \n",
    "\n",
    "p3\n",
    "\n",
    "# To file\n",
    "pdf(\"genome_fig_3_prots.pdf\")\n",
    "p3\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
